packages/calypso-e2e/src/lib/components/sidebar-component.ts

```diff
@@ -37,119 +21,107 @@ export class SidebarComponent {
-		const sidebar = await this.waitForSidebarInitialization();
-
-		// If mobile, sidebar is hidden by default and focus is on the content.
-		// The sidebar must be first brought into view.
-		if ( viewportName === 'mobile' ) {
-			await this._openMobileSidebar();
+	async navigate( item: string, subitem?: string ): Promise< void > {
+		if ( getViewportName() === 'mobile' ) {
+			await this.openMobileSidebar();
 		}
 
-		if ( item ) {
-			item = toTitleCase( item ).trim();
-			// This will exclude entries where the `heading` term matches multiple times
-			// eg. `Settings` but they are sub-headings in reality, such as Jetpack > Settings.
-			// Since the sub-headings are always hidden unless heading is selected, this works to
-			// our advantage by specifying to match only visible text.
-			selector = `${ selectors.heading } span:has-text("${ item }"):visible`;
-			await this._click( selector );
-		}
+		const itemSelector = `${ selectors.sidebar } :text-is("${ item }"):visible`;
+		await this.scrollItemIntoViewIfNeeded( itemSelector );
 
 		if ( subitem ) {
-			subitem = toTitleCase( subitem ).trim();
-			// If there is a subheading, by definition the expanded menu element will always be present.
-			await sidebar.waitForSelector( selectors.expandedMenu );
-			// Explicitly select only the child headings and combine with the text matching engine.
-			// This works better than using CSS pseudo-classes like `:has-text` or `:text-matches` for text
-			// matching.
-			selector = `${ selectors.subheading } >> text="${ subitem }"`;
-			await this._click( selector );
+			// Click top-level item without waiting for navigation if targeting subitem.
+			await this.page.click( itemSelector );
+
+			const subitemSelector = `:text-is("${ subitem }"):visible:below(${ itemSelector })`;
+			await this.scrollItemIntoViewIfNeeded( subitemSelector );
+
+			await Promise.all( [ this.page.waitForNavigation(), this.page.click( subitemSelector ) ] );
+		} else {
+			await Promise.all( [ this.page.waitForNavigation(), this.page.click( itemSelector ) ] );
 		}
 
-		// Confirm the focus is now back to the content, not the sidebar.
-		await this.page.waitForSelector( `${ selectors.layout }.focus-content` );
+		/**
+		 * Retry if the click missed the expected item. This can happen because of
+		 * the lazy-loading nature of the sidenav items.
+		 */
+		try {
+			const selectedItemSelector = `${ selectors.sidebar } .selected :text-is("${
+				subitem || item
+			}")`;
+			await this.page.waitForSelector( selectedItemSelector, {
+				timeout: 3000,
+				state: 'attached',
+			} );
+			return;
+		} catch {
+			if ( this.gotoItemRetryCount === 0 ) {
+				const itemPath = subitem ? `${ item } > ${ subitem }` : item;
+				throw new Error( `Couldn't navigate to ${ itemPath }: Expected (sub)item was not active.` );
+			}
+			this.gotoItemRetryCount -= 1;
+			await this.page.reload();
+			return this.navigate( item, subitem );
+		}
```

Comment:
This is definitely not a required comment, but is there any reason we can't just do this in a loop instead of involving class state and recursion?

In the current state, each class instance has a fixed number of retries before they are "used up". This means that if someone made one instance of this class in a test script, but did a couple of navigations using that same sidebar, the later navigations may not have retries baked in.

Doing this in a loop instead would make each navigation attempt have the same number of retries!

Reply:
Ah, great point, @dpasque! 🙌 I'm going to address this in a follow-up PR right away. Thanks!